src/Storages/StorageS3.cpp

```diff
@@ -813,10 +813,18 @@ class StorageS3Sink : public SinkToStorage
         cancelled = true;
     }
 
-    void onException() override
+    void onException(std::exception_ptr exception) override
     {
         std::lock_guard lock(cancel_mutex);
-        finalize();
+        try
+        {
+            std::rethrow_exception(exception);
+        }
+        catch(...)
+        {
+            /// An exception context is needed to proper delete write buffers without finalization
+            release();
+        }
```

Comment:
I think we should do the same for other storages that work with files like `StorageFile/StorageURL/StorageHDFS/StorageAzureBlob`.